### PR TITLE
Feat/if none match

### LINKED ISSUE #73: Task - If-None-Match header implementation for PUT/PATCH methods
In GitLab by @nrevelant on Oct 21, 2022, 11:08

### PR DESCRIPTION
**Pull request by nick-rv**
Precondition headers management factorized (including if-none-match for instance).

For information:
- a route `/etags/purge` has been added to allow to purge the cache content through the REST service. (For test purpose for instance)
- After refactoring, the handleItem() function computes now the GET/PUT/PATCH requests on a feature :
  `/collections/{id}/items/{fid}`
  `/collections/{id}/items/{fid}.{fmt}`

### CODE REVIEW

**Review by azarz — CHANGES_REQUESTED** [review]
Code factorisation needed

File: internal/data/catalog_db.go [hunk 1]
```diff
@@ -547,13 +551,29 @@ func (cat *catalogDB) CheckStrongEtags(etagsList []string) (bool, error) {
...
+		activated = true
+	}
+	if activated {
+		cache_size := cat.cache.Size()
+		cat.cache = &CacheNaive{make(map[string]interface{}, cache_size)}
+	} else {
+		cat.cache = &CacheDisabled{}
+	}
+
+	return true
```
Comment (benoitdm-oslandia) [inline internal/data/catalog_db.go]: could be replaced by a call to catalogInstance.GetCache().Reset(). You should not have Cacher implementation code outside Cacher implementations.

File: internal/data/catalog_db.go [hunk 2]
```diff
@@ -547,13 +551,29 @@ func (cat *catalogDB) CheckStrongEtags(etagsList []string) (bool, error) {
 	for _, strongEtag := range etagsList {
 		found, err := cat.cache.ContainsWeakEtag(strongEtag)
 		if err != nil {
-			return false, err
+			return true, err
```
Comment (benoitdm-oslandia) [inline internal/data/catalog_db.go]: `CheckStrongEtags` need to renamed as `IsOneEtagInCache` or something like that. Add a comment.

File: internal/service/handler.go [hunk 3]
```diff
@@ -643,37 +652,79 @@ func handleItem(w http.ResponseWriter, r *http.Request) *appError {
...
-		notModified, err := catalogInstance.CheckStrongEtags(eTagsList)
+		precondition, err := catalogInstance.CheckStrongEtags(eTagsList)
 		if err != nil {
-			return appErrorBadRequest(err, "Malformed etags are considered as not found in cache")
+			return appErrorBadRequest(err, "Malformed etags")
 		}
-		if notModified {
-			w.WriteHeader(http.StatusNotModified) // weak etag detected into the catalog cache
+		if !precondition {
+			w.WriteHeader(http.StatusNotModified) // weak etag detected into the cache
```
Comment (benoitdm-oslandia) [inline internal/service/handler.go]: I will need more explanations for this part

**Review by benoitdm-oslandia — CHANGES_REQUESTED** [review]

File: internal/service/handler.go (hunk 3 above)
Comment (nick-rv) [reply internal/service/handler.go]: This has been done this way according to the RFC guidelines which describes the precondition evaluation method : https://www.rfc-editor.org/rfc/rfc7232.html#section-3

That way i think it is easier to evaluate the implementation of the RFC and to compare with the code.

For example : 
```
If the field-value is "*", the condition is false if the origin server does
   not have a current representation for the target resource.  If the
   field-value is a list of entity-tags, the condition is false if none
   of the listed tags match the entity-tag of the selected
   representation.
```

File: internal/data/catalog_db.go (hunk 2 above)
Comment (nick-rv) [reply internal/data/catalog_db.go]: See https://github.com/benoitdm-oslandia/pg_featureserv/pull/141#discussion_r1032765938

File: internal/data/catalog_db.go (hunk 1 above)
Comment (nick-rv) [reply internal/data/catalog_db.go]: I tried many different ways to do it.
Talking about golang, we seem to face interface, structures and pointers, but not class and object instances.
I think this issue may be managed in a future PR, trying to get a uniform implementation between single instance cache and shared Redis cache, don't you think ?

**Comment by benoitdm-oslandia** [comment]
Rework, mock tests are OK. Need to review db tests

File: internal/data/listener.go [hunk 4]
```diff
@@ -136,13 +136,24 @@ func (listener *listenerDB) listenOneNotification(ctx context.Context) {
...
+		// TODO retrieve the id
+		weakEtag := api.MakeWeakEtag(collection, "", notificationData.New_xmin, api.GetCurrentHttpDate())
+		weakEtag.Data = notificationData.Data
+
+		// ===== DOUBLE ADD!!
+		_, err = listener.cache.AddWeakEtag(weakEtag.CacheKey(), weakEtag)
+		if err != nil {
+			log.Warnf("Error adding weak Etag to cache: %v", err)
+		}
+		_, err = listener.cache.AddWeakEtag(weakEtag.AlternateCacheKey(), weakEtag)
```
Comment (jmkerloch) [inline internal/data/listener.go]: @benoitdm-oslandia the feature ID is not defined so AlternateCacheKey won't return a valid key.

File: internal/service/handler.go [hunk 5]
```diff
@@ -635,170 +649,180 @@ func linksItems(name string, urlBase string) []*api.Link {
...
-
-	check, errChck := tbl.CheckTableFields(val)
-	if !check && errChck != nil {
-		return appErrorBadRequest(errChck, "validation error")
-	}
+	// Performing request according to its method
+	switch r.Method {
+	case http.MethodGet:
+		// GET
+		if checkPrecondition {
```
Comment (jmkerloch) [inline internal/service/handler.go]: @benoitdm-oslandia could be only one if

File: internal/service/handler.go [hunk 6]
```diff
@@ -840,10 +864,26 @@ func writeItemJSON(ctx context.Context, w http.ResponseWriter, tableName string,
...
-	strongEtag := fmt.Sprintf(`"%s-%d-%s-%s"`, tableName, crs, "json", feature.WeakEtag)
-	encodedStrongEtag := base64.StdEncoding.EncodeToString([]byte(strongEtag))
+	strongEtag := api.MakeStrongEtag(feature.WeakEtag.Collection, feature.WeakEtag.FeatureId, feature.WeakEtag.Etag,
+		feature.WeakEtag.LastModified, crs, "json")
+	encodedStrongEtag := strongEtag.ToEncodedString()
 	w.Header().Set("Etag", encodedStrongEtag)
-	w.Header().Set("Last-Modified", feature.LastModifiedDate)
+	w.Header().Set("Last-Modified", strongEtag.WeakEtagData.LastModified)
+
+	// TODO should not be here! What if the output format is html?
```
Comment (jmkerloch) [inline internal/service/handler.go]: @benoitdm-oslandia as the comment says, they can be no cache if response is in JSON format.

**Review by azarz — DISMISSED** [review]
LGTM, aside from jmkerloch's comments

File: internal/data/listener.go (hunk 4 above)
Comment (benoitdm-oslandia) [reply internal/data/listener.go]: fixed!

File: internal/service/handler.go (hunk 5 above)
Comment (benoitdm-oslandia) [reply internal/service/handler.go]: done!

File: internal/service/handler.go (hunk 6 above)
Comment (benoitdm-oslandia) [reply internal/service/handler.go]: done!

**Review by benoitdm-oslandia — APPROVED** [review]

**Review by azarz — APPROVED** [review]

**Pull request merged**